### PR TITLE
Di 22749 menu builder

### PR DESCRIPTION
**Pull request by michaelatias**

### CODE REVIEW

File: cpi-side/events/models/helpers/menu-builder.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { DrawnMenuBlock, ListMenu, ListMenuBlock } from "../../../configuration/models/menu.model";
+import { ListState } from "../list-state.model";
```
Comment (roiaharon-pepperi) [inline cpi-side/events/models/helpers/menu-builder.ts:2]: why is the helpers under model? why aren't them under root?

File: cpi-side/events/helpers/menu-builder.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        const drawnBlocks = await (await this.drawBlocks(menu.Blocks, state, changes)).filter(block => block) as MenuBlock[]
+        if(drawnBlocks.length == 0){
+            return null
+        }
+        return  {
+            Blocks: drawnBlocks
+        }
+
+    }
+    private async drawBlocks(menuBlocks: ListMenuBlock[], state: ListState | undefined, changes: Partial<ListState>): Promise<(MenuBlock | null)[]>{
```
Comment (roiaharon-pepperi) [inline cpi-side/events/helpers/menu-builder.ts:19]: change state to be optional (with '?'), it's more clean approach

File: cpi-side/events/helpers/menu-builder.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        return await Promise.all(menuBlocks.map(async block => {
+            return await this.callDrawBlockFunction(block, state, changes)
+        }))
+    }
+    async callDrawBlockFunction(block: ListMenuBlock, state: ListState | undefined, changes: Partial<ListState>): Promise<MenuBlock | null>{
```
Comment (roiaharon-pepperi) [inline cpi-side/events/helpers/menu-builder.ts:24]: when will this function return null? why not undefined?

File: cpi-side/test/events/helpers/menu-builder.spec.ts [hunk 4]
```diff
@@ -12,38 +13,47 @@ function getDefaultListCopy(): List{
...
 
-            const result = await stubbedMenuBuilder.build(list.Menu, defaultState, undefined)
+            const result = await stubbedMenuBuilder.build(list.Menu, defaultState, {})
             expect(result).to.eql({
                 Blocks: [
                     newButton
                 ]
             })
         })
+        it('2. menu blocks function return undefined', async function(){
```
Comment (roiaharon-pepperi) [inline cpi-side/test/events/helpers/menu-builder.spec.ts:46]: change the test name

File: cpi-side/test/events/helpers/menu-builder.spec.ts [hunk 5]
```diff
@@ -1,9 +1,10 @@
 import { expect } from 'chai'
-import { MenuBuilder } from '../../../events/models/helpers/menu-builder'
 import { defaultList, defaultState } from '../../../metadata'
-import { List } from '../../../configuration/models/list.model'
-import { ListMenuBlock } from '../../../configuration/models/menu.model'
 import * as sinon from "ts-sinon";
+import { List } from '../../../models/configuration/list.model';
```
Comment (roiaharon-pepperi) [inline cpi-side/test/events/helpers/menu-builder.spec.ts:4]: I don't see that you're running the tests on build. you should add npm script to your package.json to run tests

**Review by roiaharon-pepperi — CHANGES_REQUESTED** [review]

File: cpi-side/events/helpers/menu-builder.ts (hunk 2 above)
Comment (michaelatias) [reply cpi-side/events/helpers/menu-builder.ts:19]: with ? mark i can only put that in the end of the arguments

File: cpi-side/events/helpers/menu-builder.ts [hunk 6]
```diff
@@ -16,20 +17,42 @@ export class MenuBuilder{
...
-    private async drawBlocks(menuBlocks: ListMenuBlock[], state: ListState | undefined, changes: Partial<ListState>): Promise<(MenuBlock | null)[]>{
-        return await Promise.all(menuBlocks.map(async block => {
-            return await this.callDrawBlockFunction(block, state, changes)
+    /**
+     * this function optimize the number of calls to draw blocks, by grouping all the blocks that has the same relation
+     * the relation is one to one map with combining the addonUUID and the drawURL
+     * @param menuBlocks 
+     * @returns array of object that will tell where to call in order to draw the blocks
+     */
+    private getDrawFunctionsArray(menuBlocks: ListMenuBlock[]): {AddonUUID: string, DrawURL: string}[]{
```
Comment (roiaharon-pepperi) [inline cpi-side/events/helpers/menu-builder.ts:26]: the returned value should be exported as a type

File: cpi-side/events/helpers/menu-builder.ts [hunk 7]
```diff
@@ -7,7 +7,8 @@ export class MenuBuilder{
     constructor(){}
 
     async build(menu: ListMenu, state: ListState | undefined, changes: Partial<ListState>): Promise<Menu | null>{
-        const drawnBlocks = await (await this.drawBlocks(menu.Blocks, state, changes)).filter(block => block) as MenuBlock[]
+        const drawFunctionsArray = this.getDrawFunctionsArray(menu.Blocks)
+        const drawnBlocks = (await this.drawBlocks(drawFunctionsArray, state, changes)).filter(block => block).flat() as MenuBlock[]
```
Comment (roiaharon-pepperi) [inline cpi-side/events/helpers/menu-builder.ts:11]: why do you need the flat? draw block is returning an array

**Review by roiaharon-pepperi — CHANGES_REQUESTED** [review]

File: cpi-side/events/helpers/menu-builder.ts (hunk 7 above)
Comment (roiaharon-pepperi) [reply cpi-side/events/helpers/menu-builder.ts:12]: why would it? what am I missing?

**Pull request merged**